### PR TITLE
feat: add unique name generation to TestBase

### PR DESCRIPTION
**Pull request by prototypicalpro**
This PR reworks `TestBase#_test_name` to ensure that every test name is unique, and removes the equivalent implementations in `ShellTest` and `DeviceActionBase`. Because this generated test name is no longer consistent across runs, some tests were also reworked to only match the output values and ignore the names.

### CODE REVIEW

File: pluma/test/shelltest.py [hunk 1]
```diff
@@ -9,14 +9,11 @@
 
 
 class ShellTest(TestBase):
-    '''Execute script within the target (or host) shell'''
-    shell_test_index = 0
-
-    def __init__(self, board: Board, script: Union[str, List[str]], name: str = None,
+    def __init__(self, board: Board, script: str, name: str = None,
```
Comment (Pixep) [inline pluma/test/shelltest.py:12]: Why dropping the type hint here? `script: Union[str, List[str]]`  Typo?

Comment (Pixep) [inline pluma/test/shelltest.py:12]: Why dropping the type hint here? `script: Union[str, List[str]]` Typo?

File: pluma/test/testbase.py [hunk 2]
```diff
@@ -15,9 +16,10 @@ def __init__(self, board: Board = None, test_name_suffix: str = None):
         else:
             self._test_name = self.__class__.__name__
 
-        if test_name_suffix:
-            assert isinstance(test_name_suffix, str)
-            self._test_name += f'_{test_name_suffix}'
+        TestBase.test_count += 1
+        if test_name:
+            self._test_name += f'[{test_name}]'
+        self._test_name += f'[{TestBase.test_count}]'
```
Comment (Pixep) [inline pluma/test/testbase.py:22]: Maybe instead `Module.MyTest[abcd]#123`?

File: pluma/test/testrunner.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import re
 from abc import ABC, abstractmethod
 from copy import copy
-from typing import Iterable, Optional, List, Union
+from typing import Iterable, List, Union, Optional
```
Comment (Pixep) [inline pluma/test/testrunner.py:7]: What?! :D

File: tests/generic/test/test_TestRunner.py [hunk 4]
```diff
@@ -328,8 +329,8 @@ def teardown(self):
         def random_func_not_hook(self):
             pass
 
-    expected_data = {
-        'test_TestRunner.MyTest': {
```
Comment (Pixep) [inline tests/generic/test/test_TestRunner.py:332]: Mhh I see, if you need to remove it everywhere, do you still have a test somewhere that asserts that the test name in the data is valid?

File: tests/generic/utils.py [hunk 5]
```diff
@@ -47,3 +48,47 @@ def nonblocking(f, *args, **kwargs):
     pool = ThreadPool(processes=1)
     async_result = pool.apply_async(f, args=args, kwds=kwargs)
     return async_result
+
+
+class PlumaOutputMatcher:
```
Comment (Pixep) [inline tests/generic/utils.py:53]: Nice :)

**Review by Pixep — CHANGES_REQUESTED** [review]
Minor comments, looking good

File: pluma/test/testrunner.py (hunk 3 above)
Comment (prototypicalpro) [reply pluma/test/testrunner.py:7]: A controversial decision for sure, but one I'm willing to make.

File: pluma/test/shelltest.py (hunk 1 above)
Comment (prototypicalpro) [reply pluma/test/shelltest.py:12]: Good catch, that must've gotten messed up in the rebase.

File: tests/generic/test/test_TestRunner.py (hunk 4 above)
Comment (prototypicalpro) [reply tests/generic/test/test_TestRunner.py:332]: I added an name validation step to `PlumaOutputMatcher`, does that address the issue?

Comment (Pixep) [reply tests/generic/test/test_TestRunner.py:332]: Yes thanks. Wasn't necessary for all of these, it was very redundant (but necessary). But that works :)

File: pluma/test/shelltest.py [hunk 6]
```diff
@@ -9,14 +9,11 @@
 
 
 class ShellTest(TestBase):
-    '''Execute script within the target (or host) shell'''
```
Comment (Pixep) [inline pluma/test/shelltest.py:12]: Don't drop the class docstring :)

File: pluma/test/testbase.py [hunk 7]
```diff
@@ -15,9 +16,10 @@ def __init__(self, board: Board = None, test_name_suffix: str = None):
         else:
             self._test_name = self.__class__.__name__
 
-        if test_name_suffix:
-            assert isinstance(test_name_suffix, str)
-            self._test_name += f'_{test_name_suffix}'
+        TestBase.test_count += 1
+        if test_name:
+            self._test_name += f'[{test_name}]'
+        self._test_name += f'#{TestBase.test_count}'
```
Comment (Pixep) [inline pluma/test/testbase.py:22]: Could you cover this with a quick UT? Then we're good

**Review by Pixep — CHANGES_REQUESTED** [review]
One UT, and restoring one docstring, and good to go!

File: pluma/test/testbase.py (hunk 7 above)
Comment (prototypicalpro) [reply pluma/test/testbase.py:22]: This look good? https://github.com/Witekio/pluma-automation/blob/547cdaffeb175a2327a36de2b506b04d79fc7d7f/tests/generic/test/test_TestBase.py#L4-L21

Comment (prototypicalpro) [reply pluma/test/testbase.py:22]: Forgot `_test_name` should be accessed using `str`. How about this: https://github.com/Witekio/pluma-automation/blob/db5dee056b47becad7d107f506fa1492d4cfe0cb/tests/generic/test/test_TestBase.py#L4-L23

Comment (Pixep) [reply pluma/test/testbase.py:22]: Yep :)

**Review by Pixep — APPROVED** [review]
Thanks, all good!

**Pull request merged**